activerecord/lib/active_record/attribute_methods/serialization.rb

```diff
@@ -69,12 +69,19 @@ def serialize(attr_name, class_name_or_coder = Object)
             Coders::YAMLColumn.new(attr_name, class_name_or_coder)
           end
 
-          decorate_attribute_type(attr_name, :serialize) do |type|
-            if type_incompatible_with_serialize?(type, class_name_or_coder)
-              raise ColumnNotSerializableError.new(attr_name, type)
+          attr_name = attr_name.to_s
+          type, options = attributes_to_define_after_schema_loads[attr_name]
+
+          attribute(attr_name) do |cast_type|
+            if type && !type.is_a?(Proc)
+              cast_type = _lookup_cast_type(attr_name, type, options)
+            end
```

Comment:
It looks like this `type.is_a?(Proc)` check is here to avoid decorating with `Type::Serialized` twice, but there are now multiple ways that the type could be specified as a proc. How about checking the type directly:

```suggestion
            cast_type = _lookup_cast_type(attr_name, type, options) if type
            cast_type = cast_type.subtype if cast_type.is_a?(Type::Serialized)
```

Reply:
Originally there is no official way to decorate with `Type::Serialized` twice.

Personally I'd like to remove this part entirely, but existing tests expects `serialize` can overlay on the user-defined attribute. This part is for passing the tests.

https://github.com/rails/rails/blob/57daae230b9dc3f5166e78e0e6d103bc8faadd24/activerecord/test/cases/json_attribute_test.rb#L13-L14

https://github.com/rails/rails/blob/57daae230b9dc3f5166e78e0e6d103bc8faadd24/activerecord/test/cases/json_shared_test_cases.rb#L237-L239

Reply:
> Personally I'd like to remove this part entirely, but existing tests expects `serialize` can overlay on the user-defined attribute. This part is for passing the tests.

In other words, we need to allow `serialize` on user-defined attribute with symbol for passing tests, but I'd not like to allow `serialize` on user-defined attribute with block which is internally used feature.

Reply:
Right, my suggestion is another way to make those tests pass. I find `cast_type.is_a?(Type::Serialized)` easier to understand than `type.is_a?(Proc)`.

But it's just a preference, there's no practical benefit. I think I've delayed this PR enough. Thanks for your patience 🙇